server/src/platform/drbg/mod.rs

```diff
@@ -0,0 +1,128 @@
+/// the helper functions from the [helpers] module to implement the
+/// other methods.
+///
+/// ## Implementing [`Drbg`] for large entropy or nonce
+///
+/// Note that the required traits [`core::default::Default`] and
+/// [`core::convert::AsMut<u8>`] are not implemented for large arrays
+/// [u8; N] with N > 32. To be able to implement the traits required
+/// by SeedableRng for RNGs with such large seeds, the newtype pattern
+/// can be used:
+///
+/// ```rust
+/// const N: usize = 64;
+/// pub struct MyRngSeed(pub [u8; N]);
+///
+/// impl Default for MyRngSeed {
+///     fn default() -> MyRngSeed {
+///         MyRngSeed([0; N])
+///     }
+/// }
+///
+/// impl AsMut<[u8]> for MyRngSeed {
+///     fn as_mut(&mut self) -> &mut [u8] {
+///         &mut self.0
+///     }
+/// }
+/// ```
+///
+/// [`fill_bytes`]: Drbg::fill_bytes
+/// [`next_u32`]: Drbg::next_u32
+/// [`next_u64`]: Drbg::next_u64
+pub trait Drbg: Sized {
+    type Error: Debug;
+    /// Entropy type, which is restricted to types mutably-dereferenceable as u8 arrays
+    /// (we recommend either `[u8; N]` for some `N` or `GenericArray<u8, N>`).
+    type Entropy: Sized + Default + AsMut<[u8]>;
+    /// Nonce type, which is restricted to types mutably-dereferenceable as u8 arrays
+    /// (we recommend either `[u8; N]` for some `N` or `GenericArray<u8, N>`).
+    ///
+    /// According to 8.6.7: A nonce may be composed of one (or more) of the following
+    /// components (other components may also be appropriate):
+    ///
+    /// 1. A random value that is generated anew for each nonce, using an approved random bit
+    ///    generator.
+    /// 2. A timestamp of sufficient resolution (detail) so that it is different each time it
+    ///    is used.
+    /// 3. A monotonically increasing sequence number.
+    /// 4. A combination of a timestamp and a monotonically increasing sequence number, such
+    ///    that the sequence number is reset when and only when the timestamp changes. For
+    ///    example, a timestamp may show the date but not the time of day, so a sequence
+    ///    number is appended that will not repeat during a particular day.
+    ///
+    /// For case 1 above, the random value could be acquired from the same source and at the same time
+    /// as the entropy input. For case 2 above, the timestamp must be trusted. A trusted timestamp is
+    /// generated and signed by an entity that is trusted to provide accurate time information.
+    type Nonce: Sized + Default + AsMut<[u8]>;
+    /// Create a new DRBG using the given seed.
+    ///
+    /// All PRNG implementations should be reproducible: given a fixed seed,
+    /// the same sequence of output should be produced on all runs, library
+    /// versions and architectures (e.g. check endianness).
+    fn instantiate(
+        entropy_input: &Self::Entropy,
+        nonce: &Self::Nonce,
+        personalization_string: &[u8],
+    ) -> impl Drbg;
```

Comment:
Why not `Self` for return type?

Reply:
Also it's a bit of a rust anti-pattern to name a function like this `instantiate` instead of `new`, I think that's usually the preferred name for creation functions. Though I admit I don't know if that extends to trait impls.

Personally, it seems like this is something I'd expect not to be in the trait impl, and that the client will define as non-trait fn on their type, and that's how *they* construct it and pass it into us.

Do we really need to specify how to create a `Drbg` in the trait? Sounds a tad like an anti-pattern to me.

Reply:
> Why not Self for return type?
There was a reason, but It is now irrelevant.

> Also it's a bit of a rust anti-pattern to name a function like this instantiate instead of new,

Sadly that is NIST DRBG naming conversion (I think) but I will double check.

Reply:
I want to add one extra note about the need for this particular function.

DRBG has 2 uses to my understanding

A- getting seeded from entropy source which makes a random number generator.
B- getting seeded by a seed stored inside the tpm (which is constant) making it part of key derivation function.

in case of (A) instantiate is not really required
but for (B) it is, since the TPM may create instances of it on the fly.

Reply:
I confirm it is actually the name used all over the standards, bu I am not that much attached to the name, if you prefer `new`, we can use `new`.